File: 39.组合总和.cpp
/*
 * @lc app=leetcode.cn id=39 lang=cpp
 *
 * [39] 组合总和
 */
#include "cpp_includes.h"
// @lc code=start
class Solution {
 public:
  vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
    sort(candidates.begin(), candidates.end());
    vector<vector<int>> res;
    vector<int> path;
    combinationSum(res, path, candidates, target, 0);
    return res;
  }

  void combinationSum(vector<vector<int>>& res, vector<int>& path,
                      vector<int>& nums, int target, int index) {
    if (path.size() > 0 && target == 0) {
      res.push_back(path);
    }

    // if (index == nums.size() - 1) return;

    for (int i = index; i < nums.size(); i++) {
      if (nums[i] > target) break;
      path.push_back(nums[i]);
      combinationSum(res, path, nums, target - nums[i], i);
      path.pop_back();
    }
  }
};
// @lc code=end

File: source/39.组合总和.cs
/*
 * @lc app=leetcode.cn id=39 lang=csharp
 *
 * [39] 组合总和
 */
using csharpnaming;
// @lc code=start
public class Solution {
    public IList<IList<int>> CombinationSum(int[] candidates, int target) {
        var result = new List<IList<int>>();
        var sub = new List<int>();

        Array.Sort(candidates);
        Combination(result, sub, target, candidates, 0);
        return result;
    }

    void Combination(List<IList<int>> result, IList<int> sub, int remain, int[] candidates, int index) {
        if (index >= candidates.Length) return;
        if (remain == 0) {
            result.Add(new List<int>(sub));
            return;
        }

        for (int i = index; i < candidates.Length; i++) {
            if (remain < candidates[i]) break;
            sub.Add(candidates[i]);
            Combination(result, sub, remain-candidates[i], candidates, i);
            sub.RemoveAt(sub.Count-1);
        }
    }
}
// @lc code=end

